### PR TITLE
Fix race in credentialsProvider.setChangeListener

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
If the credentials change listeners fires twice before we enqueue the client initialization on the AsyncQueue, `syncEngine` will be null in the credentials callback. This PR re-orders our initialization to make sure that the first thing we execute on the client is always `FirestoreClient.initialize()`.

Fixes https://github.com/firebase/firebase-android-sdk/issues/875

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java [hunk 1]
```diff
@@ -127,6 +114,21 @@ public FirestoreClient(
...
+        (User user) -> {
+          if (initialized.compareAndSet(false, true)) {
+            hardAssert(!firstUser.getTask().isComplete(), "Already fulfilled first user task");
+            firstUser.setResult(user);
+          } else {
+            // this fires twoce and sync engine isnt' initialized
+            asyncQueue.enqueueAndForget(
+                () -> {
+                  Logger.debug(LOG_TAG, "Credential changed. Current user: %s", user.getUid());
+                  syncEngine.handleCredentialChange(user);
```
Comment (var-const) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java]: Optional: perhaps assert that `syncEngine` is not null? Or would this be trying too hard to protect against past problems?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java [hunk 2]
```diff
@@ -127,6 +114,21 @@ public FirestoreClient(
...
           }
         });
+
+    credentialsProvider.setChangeListener(
+        (User user) -> {
+          if (initialized.compareAndSet(false, true)) {
+            hardAssert(!firstUser.getTask().isComplete(), "Already fulfilled first user task");
+            firstUser.setResult(user);
+          } else {
+            // this fires twoce and sync engine isnt' initialized
```
Comment (MattSkala) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java]: twoce => twice, isnt' => isn't

Btw does it even make sense to leave the comment here? After the fix, sync engine should be initialized even if this fires twice.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java]: Please ignore this random comment :/ Removed.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java]: I don't think it will hurt. Done.

**Comment by wu-hui** [comment]
Will this happen to other SDKs?

**Comment by schmidt-sebastian** [comment]
It shouldn't happen on Web, but it may happen on iOS. iOS PR coming. Thanks for being vigilant!

**Pull request merged**